### PR TITLE
Add Matrix6 class

### PR DESCRIPTION
**Pull request by chapulina**
# 🎉 New feature

* Part of https://github.com/gazebosim/gz-sim/issues/1462
* Alternative to
    * https://github.com/gazebosim/gz-math/pull/442
    * https://github.com/gazebosim/gz-math/pull/447
* Builds on top of https://github.com/gazebosim/gz-math/pull/453

## Summary

Add a new `Matrix6` class to be used to store added mass (see https://github.com/gazebosim/gz-sim/issues/1462).

The class is an extension of the existing `Matrix4` class, with these differences:

* It has 2 more rows and columns
* It has the improvements that have been made to `Matrix4` on `main`:
    * #206 and #376
    * #283
    * #293 
* It doesn't have any of the APIs that don't apply to a 6x6 matrix, such as conversion to pose, affine transforms, etc.
* It has the bare minimum functionality that's needed by added mass, plus some things that were trivial to implement. Notably, it's missing `Determinant` and `Inverse`
* It has a `Submatrix` function, as requested in https://github.com/gazebosim/gz-math/pull/442#issuecomment-1175427987
* I did some future-proofing to make it easier to port this to `main`, by using `GZ_` header guards, removing redundant `ignition::math::` namespaces, etc.

## Test it

Run the added tests.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

File: include/ignition/math/Matrix6.hh [hunk 1]
```diff
@@ -0,0 +1,762 @@
...
+       return this->data[clamp(_row, IGN_ZERO_SIZE_T, IGN_FIVE_SIZE_T)][
+                         clamp(_col, IGN_ZERO_SIZE_T, IGN_FIVE_SIZE_T)];
+     }
+
+     /// \brief Get a mutable version the value at the specified row,
+     /// column index
+     /// \param[in] _row the row index. Index values are clamped to a
+     /// range of [0, 3].
+     /// \param[in] _col The column index. Index values are clamped to a
+     /// range of [0, 3].
```
Comment (scpeters) [inline include/ignition/math/Matrix6.hh:576]: [0, 5]

File: include/ignition/math/Matrix6.hh [hunk 2]
```diff
@@ -0,0 +1,762 @@
...
+
+      /// \brief Assignment operator. this = _mat
+      /// \param _mat Incoming matrix
+      /// \return itself
+      public: Matrix6<T> &operator=(const Matrix6<T> &_mat) = default;
+
+      /// \brief Multiplication assignment operator. This matrix will
+      /// become equal to this * _m2.
+      /// \param[in] _m2 Incoming matrix.
+      /// \return This matrix * _mat.
```
Comment (scpeters) [inline include/ignition/math/Matrix6.hh:292]: `_mat` -> `_m2`

**Review by scpeters — COMMENTED** [review]
I've made a few minor comments, aside from thinking that the multiplication operator would be better off with for loops, since it's so long.

Also, what do you think about adding a `SetSubMatrix(Matrix6Corner, Matrix3)` method?

File: include/ignition/math/Matrix6.hh [hunk 3]
```diff
@@ -0,0 +1,762 @@
...
+          this->data[5][3] * _m2(3, 4) +
+          this->data[5][4] * _m2(4, 4) +
+          this->data[5][5] * _m2(5, 4),
+
+          this->data[5][0] * _m2(0, 5) +
+          this->data[5][1] * _m2(1, 5) +
+          this->data[5][2] * _m2(2, 5) +
+          this->data[5][3] * _m2(3, 5) +
+          this->data[5][4] * _m2(4, 5) +
+          this->data[5][5] * _m2(5, 5));
```
[the remark this answers is not in the document]
Comment (chapulina) [reply include/ignition/math/Matrix6.hh]: I was trying to keep the diff to `Matrix4` as little as possible, so I followed the pattern there. But I agree with you that this is very error-prone. See [efb0f0d](https://github.com/gazebosim/gz-math/pull/455/commits/efb0f0d9726ec7128f9077d2e1698f86c0195585) for a more compact implementation which doesn't require allocating the matrix with zeroes and then changing each element one by one.

**Review by scpeters — APPROVED** [review]
LGTM with clean CI (there are a few lines over 80 characters)

**Pull request merged**